### PR TITLE
Ports/RP2: Allow CLK_SYS and dependants to be left running during LightSleep()

### PR DESCRIPTION
**Pull request by mungewell**
### Summary

This is a re-hash of my previous bug/pull request to the Pico SDK:
https://github.com/raspberrypi/pico-sdk/issues/1668

Now that MicroPython has pulled the clock initiation into our own code, we can implement directly.

When calling 'lightsleep()' pass the bit-fields for which clocks you want to remain on. If these are dependants of CLK_SYS then that is also kept on. The bit-fields are also use when re-initing the clocks, which are left on are then left as they are.

### Testing

I validate this on a RP Pico (RP2040). It is specific to the Clocks on the RP2040 and R2350.

Attached code will demonstrate using the FIFO and PIO to continuously flash the LED, whilst the CPU(s) are placed into 'lightsleep()'. This has the advantage of being MUCH lower power draw; normally 5V @ 19mA, then after ~10s it starts using 'lighsleep()' and the power draw drops to 5V @ 7mA - all whilst still flashing the LED.

The demo code is an extract of my larger project.

[blink_fifo_thread.zip](https://github.com/user-attachments/files/17673442/blink_fifo_thread.zip)

### CODE REVIEW

**Comment by mungewell** [comment]
The first commit is not needed, I guess I screwed up the update to current HEAD...
commit a28b9614d92018e047dfb2519cf69a2dce0c2a6b

**Review by projectgus — COMMENTED** [review]
(Gah, I wrote a long comment here and github seems to have eaten it. I'll try and write it out again...)

My mistake, I messed up strategically after merging the fix in #15301 that solved our USB glitching problem. I should have gone to pico-sdk maintainers right then and asked them to support these use cases, rather than committing MicroPython to a long term fork of this part of the Pico SDK. We're already getting reports of problems with sleep in RP2350, and the new RP chips and SDK code changes will only keep coming.

*EDIT: I did actually do this and the request https://github.com/raspberrypi/pico-sdk/issues/1746 is still open, I just have a bad memory! :laughing:*

Bottom line is that it's not viable for us to try and support this kind of low-level platform support code ourselves in the long term, and I should have thought strategically about it back in July.

I'll make a comment on your upstream issue report now and see if the pico-sdk developers are interested to re-engage on this. In the meantime, we *can* probably keep merging MicroPython-specific changes to `clocks_extra.c` for now - but I'd sleep better if we had a path to eventually upstreaming them.

File: ports/rp2/modmachine.c [hunk 1]
```diff
@@ -141,6 +143,12 @@ static void mp_machine_lightsleep(size_t n_args, const mp_obj_t *args) {
         }
     }
 
+    if (n_args >= 2)
+        req_sleep_en0 |= mp_obj_get_int(args[1]);
```
[the remark this answers is not in the document]
Comment (cpottle9) [reply ports/rp2/modmachine.c:147]: @mungewell, take a look at my github repo [RP2-PowerControl](https://github.com/cpottle9/RP2-PowerControl).

It implements code to disable RP2040 and RP2350 hardware blocks using the SLEEP_EN0/1 and WAKE_EN0/1 registers.
This allows you to turn off hardware blocks you don't need ever or turn off hardware blocks when sleeping (EG: time.sleep_ms).

The difference between machine.lightsleep() and time.sleep_ms() is lightsleep turns off one or both PLL blocks depending on whether USB is needed during lightsleep().
The PLL blocks are fairly power hungry.

Using the code in RP2-PowerControl you can turn off all the hardware blocks that lightsleep() turns off except for the PLLs.  Or  you can choose to keep hardware blocks you need turned on.

**Comment by projectgus** [comment]
@mungewell are you still interested in this PR?

**Comment by projectgus** [comment]
I'm going to close this. It looks like the new [pico-sdk pico_low_power feature](https://github.com/micropython/micropython/pull/16442#issuecomment-4879399090) has provided a way that we could expose this kind of functionality to MicroPython in a higher level way, if anyone wants to pursue this in the future.
